### PR TITLE
Attach metadata to specific fields

### PR DESCRIPTION
**Pull request by teymour-aldridge**
Due to a limitation in arrow-rs we were previously adding field metadata
to the batch schema.

## 🌟 What is the purpose of this PR?

Attaches relevant metadata to specific fields.

## 🔗 Related links

- [Asana task](https://app.asana.com/0/1199548034582004/1201892904543625/f) _(internal)_

### CODE REVIEW

File: packages/engine/lib/execution/src/runner/python/batch.py [hunk 1]
```diff
@@ -241,7 +217,7 @@ def flush_changes(self, schema, skip):
                 continue  # Not supposed to have this column in `cols`?
 
             field = schema.field(i_field)
-            if field.name in any_type_fields:
+            if "any_type" in field.metadata:
```
Comment (Alfred-Mountfield) [inline packages/engine/lib/execution/src/runner/python/batch.py:220]: Do you not need to do `and field.metadata["any_type"] == "1"` here?

File: packages/engine/lib/stateful/src/field/spec.rs [hunk 2]
```diff
@@ -25,6 +26,17 @@ impl FieldSpec {
         can_guarantee_non_null: bool,
         field_key: Option<RootFieldKey>,
     ) -> Field {
+        let metadata = {
+            let mut metadata = BTreeMap::new();
+            if self.field_type.nullable {
+                metadata.insert("nullable".to_string(), "1".to_string());
```
Comment (Alfred-Mountfield) [inline packages/engine/lib/stateful/src/field/spec.rs:32]: Do metadata values have to be strings or something?

File: packages/engine/lib/execution/src/runner/python/batch.py (hunk 1 above)
Comment (Alfred-Mountfield) [reply packages/engine/lib/execution/src/runner/python/batch.py]: No need to spend a lot of time investigating it though, just wondering if you have a sense as to why

**Review by Alfred-Mountfield — APPROVED** [review]
Looks good to me! Glad to see this done, I was annoyed when I found out our older version of Arrow didn't support it yet and we had to go with the hacky approach

File: packages/engine/lib/execution/src/runner/python/batch.py (hunk 1 above)
Comment (Alfred-Mountfield) [reply packages/engine/lib/execution/src/runner/python/batch.py]: Makes sense and I definitely agree with your decision. Thanks!

File: packages/engine/lib/stateful/src/field/spec.rs (hunk 2 above)
Comment (Alfred-Mountfield) [reply packages/engine/lib/stateful/src/field/spec.rs:32]: No worries

**Review by TimDiekmann — DISMISSED** [review]

**Comment by Alfred-Mountfield** [comment]
I haven't looked into the tests currently, but it was mentioned a while ago that the JS Implementation of Arrow (at least the version we're using) does not support metadata on the field-level. 

When we pick this PR back up we should investigate if that is the case, if it's the cause of the problem, and then how to fix it. Perhaps we need to open an issue upstream

**Pull request closed without merging**